src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,2015 @@
+                    Debug.Assert(sequence.ChildCount > 0);
+                    for (int i = 0, n = sequence.ChildCount - 1; i < n; i++)
+                    {
+                        Debug.Assert(IsEscapedMinus(sequence.ChildAt(i).Node));
+                    }
+#endif
+
+                    var last = sequence.ChildAt(sequence.ChildCount - 1).Node;
+                    if (IsEscapedMinus(last))
+                    {
+                        break;
+                    }
+
+                    return TryGetRangeComponentValueWorker(last, out ch);
+            }
+
+            ch = default;
+            return false;
+        }
+
+        private char GetCharValue(RegexToken hexText, int withBase)
+        {
+            unchecked
+            {
+                var total = 0;
+                foreach (var vc in hexText.VirtualChars)
+                {
+                    total *= withBase;
+                    total += HexValue(vc.Char);
+                }
+
+                return (char)total;
+            }
+        }
+
+        private int HexValue(char ch)
+        {
+            Debug.Assert(RegexLexer.IsHexChar(ch));
+            unchecked
+            {
+                var temp = (uint)(ch - '0');
+                if (temp <= 9)
+                {
+                    return (int)temp;
+                }
+
+                temp = (uint)(ch - 'a');
+                if (temp <= 5)
+                {
+                    return (int)(temp + 10);
+                }
+
+                temp = (uint)(ch - 'A');
+                if (temp <= 5)
+                {
+                    return (int)(temp + 10);
+                }
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        private bool HasProblem(RegexNodeOrToken component)
+        {
+            if (component.IsNode)
+            {
+                foreach (var child in component.Node)
+                {
+                    if (HasProblem(child))
+                    {
+                        return true;
+                    }
+                }
+            }
+            else
+            {
+                var token = component.Token;
+                if (token.IsMissing ||
+                    token.Diagnostics.Length > 0)
+                {
+                    return true;
+                }
+
+                foreach (var trivia in token.LeadingTrivia)
+                {
+                    if (trivia.Diagnostics.Length > 0)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private RegexExpressionNode ParseRightSideOfCharacterClassRange()
+        {
+            // Parsing the right hand side of a - is extremely strange (and most likely buggy) in
+            // the .net parser. Specifically, the .net parser will still consider itself on the
+            // right side no matter how many escaped dashes it sees.
```

Comment:
So we've taken chars between `@` and `_` inclusive, slid their values to `0` to `31`, right? I don't understand checking against `<space>` at that point... Aren't we really just checking to see if it's < 32, or is the space somehow more deeply involved?

Reply:
Or alternatively, if I'm right about the `@` to `_` thing, check directly for those bounds instead of doing the `@` subtraction and comparison to `<space>`?

Reply:
That seems reasonable.  This code was copied from the .net parser (where they care about the actual escape *value*.  i.e. they want to know that \ca has value '10'.  However, that odesn't apply to us.  So we can just reference hte actual char values.

Reply:
Or a comment explaining that is fine too. Just seemed like standalone magic with not explanation.

Reply:
Updated, and expanded tests in this area.